arangod/Replication/Syncer.cpp

```diff
@@ -367,36 +391,57 @@ Result Syncer::JobSynchronizer::waitForResponse(
     }
 
     if (_syncer->isAborted()) {
-      // clear result response
-      response.reset();
-
       std::lock_guard guard{_condition.mutex};
       _gotResponse = false;
       _response.reset();
       _res.reset();
-
-      // will result in returning TRI_ERROR_REPLICATION_APPLIER_STOPPED
-      break;
+      return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
     }
   }
-
-  return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
 }
 
 void Syncer::JobSynchronizer::request(std::function<void()> const& cb) {
+  TRI_ASSERT(cb);
+
   // by indicating that we have posted an async job, the caller
   // will block on exit until all posted jobs have finished
   if (!jobPosted()) {
     THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
   }
 
+  uint64_t id = 0;
+  {
+    // store callback
+    std::unique_lock guard{_condition.mutex};
+    id = ++_nextId;
+    _id = id;
+    _cb = cb;
```

Comment:
we don't want copy callback under lock, if we change signature of function it will be copied before

Reply:
the function is now being moved under the lock.